### ISSUE TITLE
Make webui a standalone application

### ISSUE THREAD

**Issue by deontologician** [rethinkdb/rethinkdb#3861]
Following a brief discussion on slack, we should investigate making the webui a standalone application, which the server might spawn as a subprocess for convenience.

Overall pros:

1. Remove special http handling code from the server
2. Remove http backend from the js -driver
3. Make it easier to add functionality like websockets etc
4. The UI needn't formulate its queries as reql, it could just hit a websocket endpoint for a specific query.
5. The webui could stand as a fairly extensive example of a realtime app built on rethinkdb that:
   - makes use of almost all of its capabilities
   - is architected like a user would actually architect their app (vs. special pleading that only the webui should use the http interface)
  - is kept up to date by necessity, vs. example code that bitrots over time
6. Might offer more flexibility, spinning up the webui server when necessary on whatever machine you wish, vs. needing to open the webui port on the server machine.

Overall cons:

1. Possible the server & webui could get out of sync
2. Building the server backend in node or something would add large runtime dependencies.
3. If backend is in python or something non-js it would need some extra logic to parse the queries coming in through the data explorer.

Pros to backend in python:

1. we already have a python runtime dependency (I believe), so a small server wouldn't be a large runtime dependency

Pros to a Node backend:

1. Could hack around having to parse js on the frontend for the data explorer by being a passthrough to the backend js driver.
2. The majority of our users are using node.js

**Comment by danielmewes** [comment]
Putting into backlog for now because I think there's not a sufficiently strong need for this at the moment, compared to the effort it is to change this.
Once we find that we really need additional features in the HTTP server we should reconsider this.

**Comment by larkost** [comment]
I don't believe that we have Python runtime dependency in the server. We do have that dependency for building some components, and for testing (even aside from testing the Python driver). So making it in Python would be a new dependency.

If we go down this route I would vote to build it in C/C++ with the V8 engine and whatever embeddable  http server library we want (there are lots of them). This probably would not be that difficult a task.

**Comment by marshall007** [comment]
Agreed, I mentioned this over in [#3771 (comment)](https://github.com/rethinkdb/rethinkdb/issues/3771#issuecomment-74205974). Another "pro" not mentioned here is that this would open up the possibility of making the data explorer language-agnostic in the future. In addition to what @danielmewes said, if `rethinkdbdash` becomes the official driver that would also be a good opportunity for a re-write.

**Comment by deontologician** [comment]
I think the difficulty with making the dataexplorer language agnostic is
that we'd need to create a python/ruby parser and serializer in js. Though
there may be simple parsers out there already. I do think it'd be killer if
we could do that eventually though

**Comment by neumino** [comment]
> Could hack around having to parse js on the frontend for the data explorer by being a passthrough to the backend js driver.

This doesn't quite work since most of the time you have an incomplete query, and therefore can't really call `eval` on it to get the last term.

That being said, I think it's a good thing to eventually take out with web ui from the server for the reasons mentioned before.
